### PR TITLE
Decimal various optimizations

### PR DESCRIPTION
**Pull request by lrubasze**
## Summary
- Various decimal optimizations 
- refactor `truncate()` to `checked_truncate()` with rounding mode as parameter

## Details
Optimizations list:
- Add ONE_TENTH constant
- Let `dec!` and `pdec!` macros use decimal constants for elementary values
- Remove `to_vec()` from Integer types
- Remove rotate methods from Integer types 
- Remove By trait from Integer types
- Add conversion to primitive integer types for `Decimal` and `PreciseDecimal`

## Testing
- Add missing tests for conversion from primitive types
- Add tests for conversion to primitive types

### For dApp Developers
Refactor scrypto sources to  `checked_truncate()`

### CODE REVIEW

File: radix-engine-common/src/macros.rs [hunk 1]
```diff
@@ -2,6 +2,30 @@
...
+    };
+    ("10") => {
+        $crate::math::Decimal::TEN
+    };
+    (100) => {
+        $crate::math::Decimal::ONE_HUNDRED
+    };
+    ("100") => {
+        $crate::math::Decimal::ONE_HUNDRED
+    };
```
Comment (iamyulong) [inline radix-engine-common/src/macros.rs:28]: Are we still planing to move "any string decimal parsing" to compile-time?

File: radix-engine-common/src/math/decimal.rs [hunk 2]
```diff
@@ -62,10 +64,11 @@ impl Decimal {
 
     pub const ZERO: Self = Self(I192::ZERO);
 
+    pub const ONE_HUNDREDTH: Self = Self(I192::from_digits([10_u64.pow(Decimal::SCALE - 2), 0, 0]));
```
Comment (iamyulong) [inline radix-engine-common/src/math/decimal.rs:67]: Can we test all these constants by converting it to string? I saw you've got a few but wanted to make sure we cover them all.

**Review by iamyulong — APPROVED** [review]
LGTM overall

File: radix-engine-common/src/math/decimal.rs (hunk 2 above)
Comment (lrubasze) [reply radix-engine-common/src/math/decimal.rs:67]: Done

**Pull request merged**